File: backend/app/schemas/story.py
from typing import List, Optional, Literal
from pydantic import BaseModel, Field

# TODO: replace some of the Literal types with enums for better validation and readability

class Character(BaseModel):
    name: str
    type: str = Field(..., description="Type of character: child, animal, dragon etc")
    gender: Optional[str] = Field(None, description="Character's gender")
    personality: Optional[str] = Field(None, description="Character's personality")

class StoryPrompt(BaseModel):
    age: int = Field(..., ge=1, le=12, description="Child's age in years (1-12)")
    language: Literal["english", "french"] = Field(..., description="Language")
    length: int = Field(..., ge=3, le=60, description="Number of paragraphs in the story")
    
    prompt: Optional[str] = Field(None, description="Custom prompt to guide the story")
    characters: Optional[List[Character]] = Field(None, description="List of characters")
    environment: Optional[str] = Field(None, description="Environment setting: forest, space, ocean, etc")
    theme: Optional[str] = Field(None, description="Story theme: friendship, magic, adventure, etc")
    
    tone: Optional[str] = Field(None, description="Tone: friendly, silly, adventurous, mysterious, wholesome, etc")
    conflict_type: Optional[str] = Field(None, description="Conflict type: quest, problem, villain, lost item, etc")
    ending_style: Optional[str] = Field(None, description="Ending style: happy, twist, moral, open")
    

class StoryRequest(BaseModel):
    prompt: StoryPrompt
    history: List[str] = Field(default_factory=list, description="All previously generated story paragraphs")
    choice: Optional[str] = Field(None, description="Reader's choice for the next step in the story")

    
class StoryResponse(BaseModel):
    choices: List[str] = Field(..., description="List of options for the next step of the story")
    history: List[str] = Field(..., description="Ordered list of all story steps up to this point")

File: frontend/src/types/story.ts
// Option constants for type safety and consistency
export const GENDER_OPTIONS = ['boy', 'girl', 'neutral'] as const;
export const PERSONALITY_OPTIONS = ['good', 'bad', 'neutral', 'kind', 'brave', 'helpful', 'mean', 'selfish', 'mischievous', 'cruel', 'evil', 'heroic'] as const;
export const TONE_OPTIONS = ['friendly', 'silly', 'adventurous', 'mysterious', 'wholesome'] as const;
export const CONFLICT_OPTIONS = ['quest', 'problem', 'villain', 'lost_item'] as const;
export const ENDING_OPTIONS = ['happy', 'twist', 'moral', 'open'] as const;

// Derived types from constants
export type Gender = typeof GENDER_OPTIONS[number];
export type Personality = typeof PERSONALITY_OPTIONS[number];
export type Tone = typeof TONE_OPTIONS[number];
export type ConflictType = typeof CONFLICT_OPTIONS[number];
export type EndingStyle = typeof ENDING_OPTIONS[number];

export interface Character {
  name: string;
  type: string;
  gender?: Gender;
  personality?: Personality;
}

export interface StoryPrompt {
  age: number;
  language: "english" | "french";
  length: number;
  prompt?: string;
  characters?: Character[];
  environment?: string;
  theme?: string;
  tone?: Tone;
  conflict_type?: ConflictType;
  ending_style?: EndingStyle;
}

export interface StoryRequest {
  prompt: StoryPrompt;
  history: string[];
  choice?: string;
  stage_plan?: { [key: string]: number };
}

export interface StoryResponse {
  choices: string[];
  history: string[];
  stage_plan: { [key: string]: number };
}
